// @rxweb/reactive-dynamic-forms/esm5/domain/component-viewer/component-view.js
import * as tslib_1 from "tslib";
var ComponentView = /** @class */ (function () {
    function ComponentView(_component, _viewContainerRef, _componentFactoryResolver, formControlConfig, dynamicFormBuildConfig) {
        this.formControlConfig = formControlConfig;
        this.dynamicFormBuildConfig = dynamicFormBuildConfig;
        this.component = _component;
        this.viewContainerRef = _viewContainerRef;
        this.componentFactoryResolver = _componentFactoryResolver;
    }
    ComponentView.prototype.create = function () {
        var componentFactory = this.componentFactoryResolver.resolveComponentFactory(this.component);
        this.componentRef = this.viewContainerRef.createComponent(componentFactory, this.viewContainerRef.length, this.viewContainerRef.parentInjector);
        this.setParams(this.componentRef.instance);
    };
    ComponentView.prototype.setParams = function (instance) {
        var e_1, _a;
        if (this.formControlConfig.inputs) {
            var propNames = Object.getOwnPropertyNames(this.formControlConfig.inputs);
            try {
                for (var propNames_1 = tslib_1.__values(propNames), propNames_1_1 = propNames_1.next(); !propNames_1_1.done; propNames_1_1 = propNames_1.next()) {
                    var propName = propNames_1_1.value;
                    this.overridePropAndSetValue(instance, propName, this.formControlConfig.inputs);
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (propNames_1_1 && !propNames_1_1.done && (_a = propNames_1.return)) _a.call(propNames_1);
                }
                finally { if (e_1) throw e_1.error; }
            }
        }
        instance["controlConfig"] = this.formControlConfig;
        instance["dynamicFormBuildConfig"] = this.dynamicFormBuildConfig;
    };
    ComponentView.prototype.overridePropAndSetValue = function (instance, propName, inputs) {
        var descriptor = Object.getOwnPropertyDescriptor(Object.getPrototypeOf(inputs), propName);
        var value = '';
        if (!descriptor) {
            value = inputs[propName];
            delete inputs[propName];
        }
        Object.defineProperty(inputs, propName, {
            get: function () { return (descriptor) ? descriptor.get() : value; },
            set: function (v) { instance[propName] = v; (descriptor) ? descriptor.set.call(inputs, v) : value = v; }
        });
        instance[propName] = descriptor ? descriptor.get.call(inputs) : value;
    };
    ComponentView.prototype.rootNode = function () {
        return this.componentRef.hostView.rootNodes[0];
    };
    ComponentView.prototype.getComponentRef = function () {
        return this.componentRef;
    };
    ComponentView.prototype.destroy = function () {
        if (this.componentRef) {
            this.componentRef.destroy();
            this.componentRef = undefined;
        }
    };
    return ComponentView;
}());
export { ComponentView };
