// @rxweb/reactive-dynamic-forms/esm2015/const/app.const.js
export const FILE = 'file';
export const RANGE = 'range';
export const CHECKBOX = 'checkbox';
export const BLANK = "";
export const HIDE = "hide";
export const PROP = "prop";
export const ATTR = "attr";
export const EVENTS = "events";
export const FOCUS = "focus";
export const DESCRIPTION = "description";
export const LABEL = "label";
export const READONLY = "readonly";
export const INNER_TEXT = "innertext";
export const CLASS = "class";
export const STYLE = "style";
export const RXWEB = "rxweb-";
export const RADIO = "radio";
export const CHECKED = "checked";
export const STRING = "string";
export const FUNCTION = "function";
export const SQUARE = "{";
export const COLON = ":";
export const CLICK = "click";
export const INPUT = "input";
export const TEXT = "text";
export const CONTROL = 'control';
export const BLUR = 'blur';
export const SELECT = "select";
export const TEXTAREA = "textarea";
export const BOOLEAN = "boolean";
export const NONE = "none";
export const DISPLAY = "display";
export const ADDITIONAL_CLASS = "additional-class";
export const SELECT_MULTIPLE = "select-multiple";
export const SOURCE = "source";
export const RXWEB_ID_STRING = "string";
export const ADVANCE = "advance";
export const SQUARE_CONTROL = "[control]";
export const SQUARE_LABEL = "[label]";
export const SQUARE_SMALL = "[small]";
export const SQUARE_ERROR = "[error]";
export const FILTER = "filter";
export const INPUT_TEXT = "input-text";
export const PREPEND_LEFT = "prepend-left";
export const PREPEND_RIGHT = "prepend-right";
export const PREPEND_BOTH = "prepend-both";
export const CONDITIONAL_VALIDATOR = "conditionalValidator";
export const INLINE = "inline";
