// @rxweb/reactive-dynamic-forms/esm5/domain/dom/base-object-accessor.js
import { COLON, BLANK, STRING, SQUARE } from '../../const/app.const';
import { objectPropValue } from '../../functions/object-prop-value.function';
var PROPS = ":props.";
var GLOBAL_MATCH = "g";
var DOT = ".";
var BaseObjectAccessor = /** @class */ (function () {
    function BaseObjectAccessor(dynamicNodeConfig) {
        this.dynamicNodeConfig = dynamicNodeConfig;
        this.subscribeProps = {
            names: [], props: {}
        };
        this.controlConfig = this.dynamicNodeConfig.controlConfig;
    }
    BaseObjectAccessor.prototype.getPropName = function (text) {
        if (text[0] == COLON || (text[0] == SQUARE)) {
            return text.replace(new RegExp(COLON, GLOBAL_MATCH), BLANK).replace(new RegExp(SQUARE, GLOBAL_MATCH), BLANK);
        }
        return text;
    };
    BaseObjectAccessor.prototype.getValue = function (text) {
        if (typeof text == STRING && ((text[0] == COLON) || (text[0] == SQUARE))) {
            text = text.replace(new RegExp(COLON, GLOBAL_MATCH), BLANK).replace(new RegExp(SQUARE, GLOBAL_MATCH), BLANK);
            return objectPropValue(text, this.controlConfig);
        }
        return text;
    };
    BaseObjectAccessor.prototype.setPropSubscription = function (propName, type, attributeName, valuePropName, parentPropName) {
        if (valuePropName === void 0) { valuePropName = ''; }
        if (parentPropName === void 0) { parentPropName = ''; }
        if (propName.startsWith(PROPS))
            this.defineProp(propName);
        var prop = this.getPropName(propName);
        if (!this.subscribeProps.props[prop])
            this.subscribeProps.props[prop] = {};
        if (!this.subscribeProps.props[prop][type])
            this.subscribeProps.props[prop][type] = {};
        if (parentPropName) {
            this.subscribeProps.props[prop][type][parentPropName] = {};
            this.subscribeProps.props[prop][type][parentPropName][attributeName] = (valuePropName) ? valuePropName : propName;
        }
        else
            this.subscribeProps.props[prop][type][attributeName] = (valuePropName) ? valuePropName : propName;
        if (this.subscribeProps.names.indexOf(prop) == -1)
            this.subscribeProps.names.push(prop);
    };
    BaseObjectAccessor.prototype.isSubscribeProp = function (propName) {
        return (typeof propName == STRING && (propName[0] == COLON));
    };
    BaseObjectAccessor.prototype.defineProp = function (propName) {
        var splitText = propName.split(DOT);
        if (splitText.length > 1) {
            var descriptor = Object.getOwnPropertyDescriptor(Object.getPrototypeOf(this.controlConfig.props), splitText[1]);
            if (!descriptor)
                this.controlConfig.defineProp(splitText[1]);
        }
    };
    return BaseObjectAccessor;
}());
export { BaseObjectAccessor };
