// @rxweb/reactive-dynamic-forms/esm5/const/app.const.js
export var FILE = 'file';
export var RANGE = 'range';
export var CHECKBOX = 'checkbox';
export var BLANK = "";
export var HIDE = "hide";
export var PROP = "prop";
export var ATTR = "attr";
export var EVENTS = "events";
export var FOCUS = "focus";
export var DESCRIPTION = "description";
export var LABEL = "label";
export var READONLY = "readonly";
export var INNER_TEXT = "innertext";
export var CLASS = "class";
export var STYLE = "style";
export var RXWEB = "rxweb-";
export var RADIO = "radio";
export var CHECKED = "checked";
export var STRING = "string";
export var FUNCTION = "function";
export var SQUARE = "{";
export var COLON = ":";
export var CLICK = "click";
export var INPUT = "input";
export var TEXT = "text";
export var CONTROL = 'control';
export var BLUR = 'blur';
export var SELECT = "select";
export var TEXTAREA = "textarea";
export var BOOLEAN = "boolean";
export var NONE = "none";
export var DISPLAY = "display";
export var ADDITIONAL_CLASS = "additional-class";
export var SELECT_MULTIPLE = "select-multiple";
export var SOURCE = "source";
export var RXWEB_ID_STRING = "string";
export var ADVANCE = "advance";
export var SQUARE_CONTROL = "[control]";
export var SQUARE_LABEL = "[label]";
export var SQUARE_SMALL = "[small]";
export var SQUARE_ERROR = "[error]";
export var FILTER = "filter";
export var INPUT_TEXT = "input-text";
export var PREPEND_LEFT = "prepend-left";
export var PREPEND_RIGHT = "prepend-right";
export var PREPEND_BOTH = "prepend-both";
export var CONDITIONAL_VALIDATOR = "conditionalValidator";
export var INLINE = "inline";
