// @rxweb/reactive-dynamic-forms/esm2015/services/form-control-config.js
import { BaseFormControlConfig } from './base-form-control-config';
import { dynamicContainer } from "../core/dynamicContainer";
export class FormControlConfig extends BaseFormControlConfig {
    constructor(fieldConfig, controlsConfig, notificationId) {
        super(controlsConfig, notificationId);
        this.controlsConfig = controlsConfig;
        this.overrides = {};
        this.config = fieldConfig;
        this.overrideProperty();
        this.value = fieldConfig.value;
        super.checkFilterFunction();
        this.props = this.config.props || Object.create({});
        this.setNotification();
    }
    overrideProperty() {
        if (this.config.overrideProps) {
            this.config.overrideProps.forEach(t => {
                let propInfo = dynamicContainer.getOverrideProp(t);
                this.overrideProp(propInfo);
            });
        }
    }
    set formControl(value) {
        this._formControl = value;
        setTimeout(() => this.overrideErrorsProp(this._formControl), 10);
    }
    get formControl() {
        return this._formControl;
    }
    get errorMessage() {
        return this.formControl.errorMessage;
    }
    get prependText() {
        return this.config.ui ? this.config.ui.prependText : '';
    }
}
export class ControlConfig extends FormControlConfig {
    constructor(fieldConfig, controlsConfig) {
        super(fieldConfig, controlsConfig, 0);
    }
}
