// @rxweb/reactive-dynamic-forms/esm5/services/dynamic-form-builder.js
import * as tslib_1 from "tslib";
import { RxFormGroup, RxFormArray, RxFormControl, RxwebValidators } from '@rxweb/reactive-form-validators';
import { FormControlConfig } from "./form-control-config";
import { getInstance } from "../functions/get-instance.function";
import { ApplicationUtil } from "../util/application-util";
import { NotificationState } from "../statics/control-state";
var ARRAY = "array";
var RxDynamicFormBuilder = /** @class */ (function () {
    function RxDynamicFormBuilder() {
    }
    RxDynamicFormBuilder.prototype.formGroup = function (fields, dynamicFormConfig) {
        var _this = this;
        var notificationId = NotificationState.notificationId++;
        NotificationState.notifications[notificationId] = {};
        this.formConfiguration = dynamicFormConfig || {};
        var entityObject = {};
        var formFieldConfigs = new Array();
        var modelConfig = {};
        var formGroup = new RxFormGroup({}, entityObject, {}, undefined);
        fields.forEach(function (x, i) {
            if (x.type == ARRAY) {
                _this.createFormArray(modelConfig, x, ApplicationUtil.getRootFormGroup(formGroup), entityObject, notificationId);
            }
            else {
                var splitName = x.name.split('.');
                var name_1 = x.name;
                if (splitName.length > 1) {
                    if (!entityObject[splitName[0]]) {
                        entityObject[splitName[0]] = {};
                        formGroup.addControl(splitName[0], new RxFormGroup({}, entityObject[splitName[0]], {}, undefined));
                        formGroup = formGroup.controls[splitName[0]];
                    }
                    else if (formGroup.controls[splitName[0]] != undefined && formGroup.controls[splitName[0]] instanceof RxFormGroup)
                        formGroup = formGroup.controls[splitName[0]];
                    name_1 = splitName[1];
                }
                else
                    formGroup = ApplicationUtil.getRootFormGroup(formGroup);
                var modelInstance = _this.getDynamicModelInstance(x, modelConfig, entityObject, name_1, notificationId);
                formGroup.addControl(name_1, modelInstance.formControl);
                formFieldConfigs.push(modelInstance);
            }
        });
        if (this.formConfiguration.additionalConfig)
            this.formConfiguration.additionalConfig.forEach(function (t) { return _this.getModelInstance(t, modelConfig, notificationId); });
        var rootFormGroup = ApplicationUtil.getRootFormGroup(formGroup);
        rootFormGroup["model"] = undefined;
        return {
            controlsConfig: modelConfig,
            formGroup: rootFormGroup
        };
    };
    RxDynamicFormBuilder.prototype.createFormArray = function (modelConfig, field, formGroup, entityObject, notificationId) {
        var _this = this;
        modelConfig[field.name] = [];
        entityObject[field.name] = [];
        var formArray = new RxFormArray(entityObject[field.name], []);
        if (field.controlConfigs) {
            if (field.rows)
                field.rows.forEach(function (row) {
                    formArray.controls.push(_this.createDynamicFormGroup(field, modelConfig[field.name], _this.getRefObject(entityObject[field.name]), row, notificationId));
                });
            if (field.minimumRepeatCount && field.minimumRepeatCount > 0) {
                var countLeft = field.minimumRepeatCount - (formArray.controls.length);
                for (var i = 0; i < countLeft; i++)
                    formArray.controls.push(this.createDynamicFormGroup(field, modelConfig[field.name], this.getRefObject(entityObject[field.name]), { fields: [] }, notificationId));
            }
            this.addTwoProp(modelConfig[field.name], field, entityObject[field.name], formArray, notificationId);
            formGroup.addControl(field.name, formArray);
        }
    };
    RxDynamicFormBuilder.prototype.getRefObject = function (entityObject) {
        var jObject = {};
        entityObject.push(jObject);
        return jObject;
    };
    RxDynamicFormBuilder.prototype.addTwoProp = function (modelConfig, x, entityObject, formArray, notificationId) {
        var _this = this;
        modelConfig.__proto__.addItem = function () {
            formArray.controls.push(_this.createDynamicFormGroup(x, modelConfig, _this.getRefObject(entityObject), { fields: [] }, notificationId));
        };
        modelConfig.__proto__.removeItem = function (index) {
            formArray.removeAt(index);
            modelConfig.splice(index, 1);
        };
    };
    RxDynamicFormBuilder.prototype.createDynamicFormGroup = function (x, modelConfig, entityObject, row, notificationId) {
        var _this = this;
        var nestedFormGroup = new RxFormGroup({}, entityObject, {}, undefined);
        var jObject = {};
        modelConfig.push(jObject);
        Object.keys(x.controlConfigs).forEach(function (key) {
            var field = row.fields.filter(function (x) { return x.name == key; })[0];
            var formControlConfig = tslib_1.__assign({}, x.controlConfigs[key], { name: key });
            if (field)
                formControlConfig = tslib_1.__assign({}, formControlConfig, field);
            var modelInstance = _this.getDynamicModelInstance(formControlConfig, jObject, entityObject, key, notificationId);
            nestedFormGroup.addControl(key, modelInstance.formControl);
        });
        return nestedFormGroup;
    };
    RxDynamicFormBuilder.prototype.getModelInstance = function (x, modelConfig, notificationId) {
        var configModel = (x.modelName) && this.formConfiguration && this.formConfiguration.controlConfigModels ? this.formConfiguration.controlConfigModels.filter(function (y) { return y.modelName == x.modelName; })[0] : undefined;
        var modelArguments = [x, modelConfig, notificationId];
        var model = undefined;
        if (configModel) {
            model = configModel.model;
            if (configModel.arguments)
                configModel.arguments.forEach(function (t) { return modelArguments.push(t); });
        }
        else
            model = FormControlConfig;
        var modelInstance = getInstance(model, modelArguments);
        modelConfig[x.name] = modelInstance;
        return modelInstance;
    };
    RxDynamicFormBuilder.prototype.getDynamicModelInstance = function (x, modelConfig, entityObject, name, notificationId) {
        var modelInstance = this.getModelInstance(x, modelConfig, notificationId);
        var validators = [];
        var asyncValidators = [];
        if (x.validators)
            this.validatorBindings(validators, x.validators);
        if (modelInstance.validator)
            validators.push(modelInstance.validator.bind(modelInstance));
        if (modelInstance.asyncValidator)
            asyncValidators.push(modelInstance.asyncValidator.bind(modelInstance));
        if (modelInstance)
            entityObject[x.name] = x.value;
        var baseObject = {};
        baseObject[x.name] = x.value;
        entityObject[x.name] = x.value;
        modelInstance.formControl = new RxFormControl(x.value, validators, asyncValidators, entityObject, baseObject, name, undefined);
        return modelInstance;
    };
    RxDynamicFormBuilder.prototype.validatorBindings = function (validations, validationConfig) {
        for (var column in RxwebValidators) {
            if (validationConfig[column]) {
                validations.push(RxwebValidators[column](validationConfig[column]));
            }
        }
        return validations;
    };
    return RxDynamicFormBuilder;
}());
export { RxDynamicFormBuilder };
