// @rxweb/reactive-dynamic-forms/esm5/services/prop-descriptor.js
import * as tslib_1 from "tslib";
import { ValueChangeNotification } from "./value-change-notification";
var FILTER = "filter";
var FUNCTION = "function";
var ERRORS = "errors";
var PropDescriptor = /** @class */ (function (_super) {
    tslib_1.__extends(PropDescriptor, _super);
    function PropDescriptor() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    PropDescriptor.prototype.checkFilterFunction = function () {
        var descriptor = Object.getOwnPropertyDescriptor(Object.getPrototypeOf(this), FILTER);
        this.isDefinedFilter = (descriptor && !descriptor.get && !descriptor.set && typeof this[FILTER] == FUNCTION);
    };
    PropDescriptor.prototype.overrideProp = function (props) {
        var _this = this;
        Object.keys(props).forEach(function (t) {
            Object.defineProperty(_this, t, props[t]);
        });
    };
    PropDescriptor.prototype.defineProp = function (propName) {
        var _this = this;
        var value = this.props[propName];
        var oldValue = null;
        Object.defineProperty(this.props, propName, {
            get: function () { return value; },
            set: function (v) {
                value = v;
                _this.notifyValueChanged("props." + propName, value, oldValue, false);
                oldValue = value;
            }
        });
    };
    PropDescriptor.prototype.overrideProps = function () {
        var _this = this;
        ["disabled", "label", "placeholder", "hide", "description", "focus", "readonly", "class", "source"].forEach(function (t) {
            var descriptor = _this.getDescriptor(t);
            var value = null;
            var oldValue = null;
            Object.defineProperty(_this, t, {
                get: function () { return descriptor ? descriptor.get.call(_this) : value; },
                set: function (v) {
                    value = v;
                    if (descriptor && descriptor.set)
                        descriptor.set.call(_this, v);
                    _super.prototype.notifyValueChanged.call(_this, t, value, oldValue);
                    oldValue = v;
                }
            });
        });
    };
    PropDescriptor.prototype.overrideErrorsProp = function (formControl) {
        var _this = this;
        var value = formControl.errors;
        var errorMessage = '';
        var oldValue = formControl.errorMessage;
        Object.defineProperty(formControl, ERRORS, {
            get: function () { return value; },
            set: function (v) {
                value = v;
                errorMessage = formControl.errorMessage;
                _this.notifyValueChanged("errorMessage", errorMessage || "", oldValue);
                oldValue = formControl.errorMessage;
            }
        });
        formControl[ERRORS] = formControl[ERRORS];
    };
    PropDescriptor.prototype.getDescriptor = function (propName) {
        var descriptor = Object.getOwnPropertyDescriptor(Object.getPrototypeOf(this), propName);
        if (!descriptor)
            descriptor = Object.getOwnPropertyDescriptor(this, propName);
        return descriptor;
    };
    return PropDescriptor;
}(ValueChangeNotification));
export { PropDescriptor };
