// @rxweb/reactive-dynamic-forms/esm5/services/value-change-notification.js
import { isNotMatched } from "../util/is-not-matched";
var ValueChangeNotification = /** @class */ (function () {
    function ValueChangeNotification() {
        this.attributeChangeSubscriptions = new Array();
    }
    ValueChangeNotification.prototype.onPropValueChanged = function (controlId, subscription, func) {
        this.attributeChangeSubscriptions.push({ controlId: controlId, names: subscription.names, props: subscription.props, func: func });
    };
    ValueChangeNotification.prototype.notifyValueChanged = function (name, value, oldValue, isProps) {
        if (isProps === void 0) { isProps = false; }
        if ((!isProps && this.isNotEqual(oldValue, value)) && this.onPropValueChanged) {
            var subscriptions = this.attributeChangeSubscriptions.filter(function (t) { return t.names.indexOf(name) != -1; });
            subscriptions.forEach(function (subscribe) {
                if (subscribe.props && subscribe.props[name])
                    subscribe.func(subscribe.props[name]);
            });
        }
    };
    ValueChangeNotification.prototype.isNotEqual = function (leftValue, rightValue) {
        if (Array.isArray(leftValue) && Array.isArray(rightValue)) {
            var isNotEqual = leftValue.length != rightValue.length;
            if (!isNotEqual)
                for (var i = 0; i < leftValue.length; i++) {
                    isNotEqual = isNotMatched(leftValue[i], rightValue[i]);
                    if (isNotEqual)
                        break;
                }
            return isNotEqual;
        }
        return leftValue != rightValue;
    };
    ValueChangeNotification.prototype.destroy = function (controlId) {
        for (var i = 0; i < this.attributeChangeSubscriptions.length; i++) {
            if (this.attributeChangeSubscriptions[i].controlId == controlId) {
                this.attributeChangeSubscriptions.splice(i, 1);
                break;
            }
        }
    };
    return ValueChangeNotification;
}());
export { ValueChangeNotification };
